### PR TITLE
CTAS failure during commit leaves table behind

### LINKED ISSUE #2537: CTAS failure during commit leaves table behind, cannot be retried
**Given**

```
presto> USE hive.default;

presto> CREATE TABLE nation AS SELECT * FROM tpch.tiny.nation;
```

**When**
```
presto> SET SESSION query_max_execution_time = '5s';
presto> CREATE TABLE t WITH (partitioned_by=ARRAY['comment']) AS SELECT * FROM nation;

Query 20200117_094529_00005_nfxfs, FAILED, 1 node
http://localhost:8080/ui/query.html?20200117_094529_00005_nfxfs
Splits: 35 total, 35 done (100.00%)
CPU Time: 0.4s total,    63 rows/s, 3.76KB/s, 20% active
Per Node: 0.1 parallelism,     4 rows/s,   255B/s
Parallelism: 0.1
Peak Memory: 0B
0:06 [25 rows, 1.49KB] [4 rows/s, 255B/s]

Query 20200117_094529_00005_nfxfs failed: Query exceeded the maximum execution time limit of 5.00s
io.prestosql.spi.PrestoException: Query exceeded the maximum execution time limit of 5.00s
	at io.prestosql.execution.QueryTracker.enforceTimeLimits(QueryTracker.java:184)
	at io.prestosql.execution.QueryTracker.lambda$start$0(QueryTracker.java:88)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Yet**
```
presto:default> SHOW TABLES;
 Table
--------
 nation
 t
(2 rows)
```

**Expected**
* query cancels (and leaves nothing behind) or query succeeds

**Notes**
The query writes data fast (less than 5s), but it takes a lot time to commit transaction, as this includes quite a few calls to the metastore.
Query gets killed after 5s run time, yet commit completes.

Query cannot be tried, because the table already exists.

### PR DESCRIPTION
**Pull request by wendigo**
Fixes #2537

First commit refactors ```QueryStateMachine``` so internally it keeps ```queryState``` as ```StateMachine<QueryStateHolder>``` (```QueryStateHolder``` wraps ```QueryState``` and ```Optional<ExecutionFailureInfo>``` (when ```QueryState``` == ```FAILED```)).

Second commit changes ```QueryStateMachine``` to disallow failing/cancelling query while it's finishing. 

Third commit fixes unnecessary call to ```transitionToFailed``` (pollutes debug log)

Forth commit refactors ```StateMachine``` so internally it tests if ```T currentState``` is terminal with a ```Predicate<T>```

And finally fifth commit uses refactored ```StateMachine``` in rest of the codebase

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/execution/QueryStateMachine.java [hunk 1]
```diff
@@ -902,7 +891,7 @@ public void addQueryInfoStateChangeListener(StateChangeListener<QueryInfo> state
 
     public ListenableFuture<QueryState> getStateChange(QueryState currentState)
     {
-        return queryState.getStateChange(currentState);
+        return Futures.transform(queryState.getStateChange(new QueryStateHolder(currentState)), QueryStateHolder::getState, directExecutor());
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/execution/QueryStateMachine.java]: For `currentState=FAILED` `QueryStateHolder::new` should throw, as this creates inconsistent state.
But FAILED is a terminal state, so you can handle this here:
```
if (currentState.isDone()) {
  verify(queryState.get().getState() == currentState, "Invalid currentState");
  return immediateFuture(currentState);
}
...
```

File: presto-main/src/main/java/io/prestosql/execution/QueryStateHolder.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+package io.prestosql.execution;
+
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static io.prestosql.util.Failures.toFailure;
+import static java.util.Objects.requireNonNull;
+
+public class QueryStateHolder
```
[the remark this answers is not in the document]
Comment (findepi) [reply presto-main/src/main/java/io/prestosql/execution/QueryStateHolder.java]: I want this class to be impl detail of `QueryStateMachine`. I'd put it as an inner static class there, at the bottom.

[the remark this answers is not in the document]
Comment (wendigo) [reply presto-main/src/main/java/io/prestosql/execution/QueryStateHolder.java]: OK.

File: presto-main/src/main/java/io/prestosql/execution/QueryStateHolder.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    @Override
+    public boolean equals(Object obj)
+    {
+        if (obj instanceof QueryStateHolder) {
+            return state.equals(((QueryStateHolder) obj).state);
+        }
+
+        return state.equals(obj);
```
[the remark this answers is not in the document]
Comment (wendigo) [reply presto-main/src/main/java/io/prestosql/execution/QueryStateHolder.java]: Yep, I've made it on purpose (so that terminalStates.contains() would work). I've changed StateMachine so it can take a Predicate<T> and now equals is not required).

File: presto-main/src/main/java/io/prestosql/execution/QueryStateHolder.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    public boolean isDone()
+    {
+        return state.isDone();
+    }
+
+    @Override
+    public int hashCode()
```
[the remark this answers is not in the document]
Comment (wendigo) [reply presto-main/src/main/java/io/prestosql/execution/QueryStateHolder.java]: Deleted hashCode

File: presto-main/src/main/java/io/prestosql/execution/QueryStateHolder.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+    public static QueryStateHolder failedWithThrowable(Throwable throwable)
+    {
+        requireNonNull(throwable, "throwable is null");
+        return new QueryStateHolder(toFailure(throwable));
+    }
+
+    public static Set<QueryStateHolder> terminalStates()
+    {
+        return QueryState.TERMINAL_QUERY_STATES.stream().map(QueryStateHolder::new).collect(toImmutableSet());
```
[the remark this answers is not in the document]
Comment (wendigo) [reply presto-main/src/main/java/io/prestosql/execution/QueryStateHolder.java]: Removed this method. Not needed anymore.

File: presto-main/src/main/java/io/prestosql/dispatcher/LocalDispatchQuery.java [hunk 6]
```diff
@@ -74,7 +74,15 @@ public LocalDispatchQuery(
...
 
-        addExceptionCallback(queryExecutionFuture, stateMachine::transitionToFailed);
+        // Check in the exception callback if we are already in the terminal state
+        // so that throwable could be ignored (CancellationException will be thrown
+        // when we cancel queryExecutionFuture in addStateChangeListener).
+        addExceptionCallback(queryExecutionFuture, throwable -> {
+            if (!stateMachine.isDone()) {
+                stateMachine.transitionToFailed(throwable);
+            }
+        });
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/dispatcher/LocalDispatchQuery.java]: This LGTM. Please move to a separate PR

File: presto-main/src/main/java/io/prestosql/execution/StateMachine.java [hunk 7]
```diff
@@ -83,7 +83,23 @@ public StateMachine(String name, Executor executor, T initialState, Iterable<T>
         this.name = requireNonNull(name, "name is null");
         this.executor = requireNonNull(executor, "executor is null");
         this.state = requireNonNull(initialState, "initialState is null");
-        this.terminalStates = ImmutableSet.copyOf(requireNonNull(terminalStates, "terminalStates is null"));
+        this.isTerminalState = state -> ImmutableSet.of(terminalStates).contains(state);
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/execution/StateMachine.java]: Make this ctor delegate to the other ctor.

Also, `ImmutableSet.of(terminalStates)::contains`

File: presto-main/src/main/java/io/prestosql/execution/StateMachine.java [hunk 8]
```diff
@@ -48,7 +47,8 @@
     private final String name;
     private final Executor executor;
     private final Object lock = new Object();
-    private final Set<T> terminalStates;
+
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/execution/StateMachine.java]: do not add a blank line here

File: presto-main/src/main/java/io/prestosql/execution/QueryStateMachine.java [hunk 9]
```diff
@@ -798,62 +787,59 @@ private void transitionToFinished()
...
-            });
-        }
-        else {
-            QUERY_STATE_LOG.debug(throwable, "Failure after query %s finished", queryId);
-        }
+        do {
+            QueryStateWrapper observedState = queryState.get();
+
+            if (observedState.isDone()) {
+                QUERY_STATE_LOG.debug(throwable, "Query %s tried to fail while %s", queryId, observedState.unwrap());
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/execution/QueryStateMachine.java]:
```suggestion
                QUERY_STATE_LOG.debug(throwable, "Failure after query %s finished", queryId);
```

File: presto-main/src/main/java/io/prestosql/execution/QueryStateMachine.java [hunk 10]
```diff
@@ -798,62 +787,59 @@ private void transitionToFinished()
...
+                return false;
+            }
 
-        return failed;
+            // When in FINISHING state allow to transition to FAILED state
+            // only when called from transitionToFinishing()
+            if (observedState.is(FINISHING) && !finishingFailed) {
+                QUERY_STATE_LOG.debug(throwable, "Query %s could not fail while FINISHING", queryId);
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/execution/QueryStateMachine.java]:
```suggestion
                QUERY_STATE_LOG.debug(throwable, "Cannot fail query %s when it is finishing", queryId);
```

File: presto-main/src/main/java/io/prestosql/execution/QueryStateMachine.java [hunk 11]
```diff
@@ -798,62 +787,59 @@ private void transitionToFinished()
...
+                return false;
+            }
+
+            if (queryState.compareAndSet(observedState, QueryStateWrapper.failedWithThrowable(throwable))) {
+                // We will mark end of query when transitioned to FAILED
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/execution/QueryStateMachine.java]: i think the comment is redundant and can be removed.

File: presto-main/src/main/java/io/prestosql/execution/QueryStateMachine.java [hunk 12]
```diff
@@ -798,62 +787,59 @@ private void transitionToFinished()
...
-            session.getTransactionId().ifPresent(transactionId -> {
-                if (transactionManager.isAutoCommit(transactionId)) {
-                    transactionManager.asyncAbort(transactionId);
-                }
-                else {
-                    transactionManager.fail(transactionId);
-                }
-            });
-        }
+        return transitionToFailed(new PrestoException(USER_CANCELED, "Query was canceled"));
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/execution/QueryStateMachine.java]: make this a separate, preparatory commit.

File: presto-main/src/main/java/io/prestosql/execution/StageStateMachine.java [hunk 13]
```diff
@@ -107,7 +106,7 @@ public StageStateMachine(
         this.tables = ImmutableMap.copyOf(requireNonNull(tables, "tables is null"));
         this.scheduledStats = requireNonNull(schedulerStats, "schedulerStats is null");
 
-        stageState = new StateMachine<>("stage " + stageId, executor, PLANNED, TERMINAL_STAGE_STATES);
+        stageState = new StateMachine<>("stage " + stageId, executor, PLANNED, StageState::isDone);
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/execution/StageStateMachine.java]: This is not related to QueryStateMachine changes.
In particular, the QueryStateMachine commit fixes a bug, but this change is just a refactor. They shouldn't be together in one commit.

Extract a preparatory commit with StateMachine change (accepting Predicate) and apply it here.
Then the rest.

File: presto-main/src/main/java/io/prestosql/execution/TaskStateMachine.java [hunk 14]
```diff
@@ -43,15 +42,8 @@
...
-        taskState.addStateChangeListener(new StateChangeListener<TaskState>()
-        {
-            @Override
-            public void stateChanged(TaskState newState)
-            {
-                log.debug("Task %s is %s", taskId, newState);
-            }
-        });
+        taskState = new StateMachine<>("task " + taskId, executor, TaskState.RUNNING, TaskState::isDone);
+        taskState.addStateChangeListener(newState -> log.debug("Task %s is %s", taskId, newState));
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/execution/TaskStateMachine.java]: Cool formatting improvement. Please separate it out as a separate commit.

**Comment by findepi** [comment]
@wendigo thanks for the update here.

Please make sure the commit history is the way you want it to be.
In particular, every commit should leave the code in the reasonable shape ("all tests pass").
Therefore the predicate in the StateMachine needs to go before introducing a composite
class for QueryState + failure cause.

File: presto-main/src/main/java/io/prestosql/dispatcher/LocalDispatchQuery.java (hunk 6 above)
Comment (wendigo) [reply presto-main/src/main/java/io/prestosql/dispatcher/LocalDispatchQuery.java]: Sure

File: presto-main/src/main/java/io/prestosql/execution/StateMachine.java (hunk 7 above)
Comment (wendigo) [reply presto-main/src/main/java/io/prestosql/execution/StateMachine.java]: It's already so at the next commit :)

File: presto-main/src/main/java/io/prestosql/execution/QueryStateMachine.java (hunk 9 above)
Comment (wendigo) [reply presto-main/src/main/java/io/prestosql/execution/QueryStateMachine.java]: I tried to keep every log line beginning the same way (Query ID ...). Does it make sense?

File: presto-main/src/main/java/io/prestosql/dispatcher/LocalDispatchQuery.java (hunk 6 above)
Comment (wendigo) [reply presto-main/src/main/java/io/prestosql/dispatcher/LocalDispatchQuery.java]: https://github.com/prestosql/presto/pull/2607
